commands2/ramsetecommand.py

```diff
@@ -0,0 +1,186 @@
+        provided trajectory. PID control and feedforward are handled internally, and the outputs are scaled
+        from -12 to 12 representing units of volts. If any one of the optional parameters are provided, each
+        other optional parameter becomes required.
+        :param feedforward      The feedforward to use for the drive
+        :param leftController:  The PIDController for the left side of the robot drive.
+        :param rightController: The PIDController for the right side of the robot drive.
+        :param wheelSpeeds:     A function that supplies the speeds of the left and right sides of the robot
+                                drive.
+        """
+        super().__init__()
+
+        self.timer = Timer()
+
+        # Store all the requirements that are required
+        self.trajectory: Trajectory = trajectory
+        self.pose = pose
+        self.follower: RamseteController = controller
+        self.kinematics = kinematics
+        self.output: Callable[[float, float], None] = outputVolts
+        self.leftController = None
+        self.rightController = None
+        self.feedforward = None
+        self.wheelspeeds = None
+        self.usePID = False
+        # Optional requirements checks. If any one of the optionl parameters are not None, then all the optional
+        # requirements become required.
+        if feedforward or leftController or rightController or wheelSpeeds is not None:
+            if feedforward or leftController or rightController or wheelSpeeds is None:
+                raise RuntimeError(
+                    f"Failed to instantiate RamseteCommand, not all optional arguments were provided.\n \
+                                    Feedforward - {feedforward}, LeftController - {leftController}, RightController - {rightController}, WheelSpeeds - {wheelSpeeds} "
+                )
+
+            self.leftController: PIDController = leftController
+            self.rightController: PIDController = rightController
+            self.wheelspeeds: Callable[[], DifferentialDriveWheelSpeeds] = wheelSpeeds
+            self.feedforward: SimpleMotorFeedforwardMeters = feedforward
+            self.usePID = True
+        self.prevSpeeds = DifferentialDriveWheelSpeeds()
+        self.prevTime = -1.0
+
+        self.addRequirements(*requirements)
+
+    def initialize(self):
+        self.prevTime = -1
+        initial_state = self.trajectory.sample(0)
+        initial_speeds = self.kinematics.toWheelSpeeds(
+            ChassisSpeeds(
+                initial_state.velocity,
+                0,
+                initial_state.curvature * initial_state.velocity,
+            )
+        )
+        self.prevSpeeds = initial_speeds
+        self.timer.restart()
+        if self.usePID:
+            self.leftController.reset()
+            self.rightController.reset()
+
+    def execute(self):
+        cur_time = self.timer.get()
+        dt = cur_time - self.prevTime
```

Comment:
Mixing camelCase and underscore_case is weird in the same file, even though I do it all the time. `curTime` instead.